internal/backup/csi_volumesnapshot.go

```diff
@@ -0,0 +1,120 @@
+limitations under the License.
+*/
+
+package backup
+
+import (
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	snapshotv1beta1api "github.com/kubernetes-csi/external-snapshotter/v2/pkg/apis/volumesnapshot/v1beta1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/vmware-tanzu/velero-plugin-for-csi/internal/util"
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/plugin/velero"
+)
+
+// VolumeSnapshotBackupItemAction is a backup item action plugin to backup
+// CSI VolumeSnapshot objects using Velero
+type VolumeSnapshotBackupItemAction struct {
+	Log logrus.FieldLogger
+}
+
+// AppliesTo returns information indicating that the VolumeSnapshotBackupItemAction should be invoked to backup volumesnapshots.
+func (p *VolumeSnapshotBackupItemAction) AppliesTo() (velero.ResourceSelector, error) {
+	p.Log.Info("VolumeSnapshotBackupItemAction AppliesTo")
+
+	return velero.ResourceSelector{
+		IncludedResources: []string{"volumesnapshots.snapshot.storage.k8s.io"},
+	}, nil
+}
+
+func setVolumeSnapshotAnnotationsAndLabels(vs *snapshotv1beta1api.VolumeSnapshot, vals map[string]string) {
+	if vs.Annotations == nil {
+		vs.Annotations = make(map[string]string)
+	}
+	if vs.Labels == nil {
+		vs.Labels = make(map[string]string)
+	}
+
+	for k, v := range vals {
+		vs.Annotations[k] = v
+		vs.Labels[k] = v
+	}
+}
+
+// Execute backs VolumeSnapshotContents associated with the volumesnapshot as additional resource.
+func (p *VolumeSnapshotBackupItemAction) Execute(item runtime.Unstructured, backup *velerov1api.Backup) (runtime.Unstructured, []velero.ResourceIdentifier, error) {
+	p.Log.Infof("Executing VolumeSnapshotBackupItemAction")
+
+	var vs snapshotv1beta1api.VolumeSnapshot
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(item.UnstructuredContent(), &vs); err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+
+	_, snapshotClient, err := util.GetClients()
+	if err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+	p.Log.Infof("Getting VolumesnapshotContent for Volumesnapshot %s/%s", vs.Namespace, vs.Name)
+	vsc, err := util.GetVolumeSnapshotContentForVolumeSnapshot(&vs, snapshotClient.SnapshotV1beta1(), p.Log)
+	if err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+
+	vals := map[string]string{
+		util.VolumeSnapshotHandleLabel: *vsc.Status.SnapshotHandle,
+		util.CSIDriverNameLabel:        vsc.Spec.Driver,
+	}
+
+	additionalItems := []velero.ResourceIdentifier{}
+
+	// Capture the vsc's deletesnapshot secret annotations into the backup of the volumesnapshot
+	if util.IsVolumeSnapshotContentHasDeleteSecret(vsc) {
+		deleteSnapshotSecretName := vsc.Annotations[util.PrefixedSnapshotterSecretNameKey]
+		deleteSnapshotSecretNamespace := vsc.Annotations[util.PrefixedSnapshotterSecretNamespaceKey]
+		// TODO: add GroupResource for secret into kuberesource
```

Comment:
This was merged, though this repo's deps will have to be set to Velero's `master` if it's not already.

Reply:
I think waiting for a release to be better as it gives us more control over the changes with which this is built once this becomes more widely used. Updating the modules to use `master` may inadvertent build breaks or can result in unexpected behavior.

Reply:
This will be in the imminent 1.3.2 release. Will fix up when updating the dependency